### ISSUE TITLE
Unexpected token D in JSON at position 1

### ISSUE THREAD

**Issue by dossy** [wongjn/vscode-php-sniffer#29]
When PHP Sniffer runs, VSCode pops up the following notification:

"Unexpected token D in JSON at position 1"

**To Reproduce**
1. Open a PHP file.
2. Save it.
3. See error

**Expected behavior**

PHP Sniffer runs as expected, no errors displayed.

**Environment**
 - OS: MacOS X 10.13.6
 - VSCode version: 1.42.1
 - PHP CLI version: 7.3.6
-  squizlabs/PHP_CodeSniffer version: 3.5.2

**Extension settings**
```json
{
    "phpSniffer.run": "onSave",
    "phpSniffer.onTypeDelay": 250,
    "phpSniffer.executablesFolder": "",
    "phpSniffer.standard": "",
    "phpSniffer.snippetExcludeSniffs" : [],
}
```

**Comment by wongjn** [comment]
Could you show me any errors in developer tools console (Help → Toggle Developer Tools → Console tab)?

**Comment by dossy** [comment]
Here's just installing the extension in VSCode, from the Developer Tools console:

![image](https://user-images.githubusercontent.com/65378/75271177-c72ed680-57c9-11ea-8959-a709efadef04.png)

**Comment by wongjn** [comment]
Are you certain that the error is coming from this plugin? (I see you have errors from `persoderlind.vscode-phpcbf` there, try launching with all plugins disabled apart from this one)
Do you have `PHP_CodeSniffer` installed globally?
Do you have `phpSniffer.autoDetect` set to `true`?

**Comment by dossy** [comment]
So, when I don't have `wongjn.php-sniffer` installed or enabled in VSCode, I don't get this error.  When I install it and enable it, the error is displayed, even on install.

`phpSniffer.autoDetect` is unset, so whatever the default is set to is what it's set to.

I have `PHP_CodeSniffer` installed globally through Homebrew in `/usr/local/bin/phpcs` and `/usr/local/bin/phpcbf`.

As a test, I disabled all my installed extensions, and then restarted VSCode, and only installed `wongjn.php-sniffer`.  I then opened a PHP file, and upon opening the file, the error toast appeared.  Here is the error in the Console:

![image](https://user-images.githubusercontent.com/65378/75283612-3237d780-57e1-11ea-85cb-5b5150821d6f.png)

This was on file open, not save or even typing into the file, yet.

**Comment by wongjn** [comment]
I can't seem to replicate this issue.

What output do you get (where `index.php` is a file that exists for you) with these commands:
- `phpcs index.php`
- `phpcs --report=json index.php`
- `phpcs --report=json --standard=PSR2 index.php`

**Comment by dossy** [comment]
* `phpcs index.php`

```
E 1 / 1 (100%)



FILE: /Users/dossy/[redacted]/index.php
-----------------------------------------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------------------------------------------
 14 | ERROR | [x] Expected 1 spaces after opening bracket; 0 found
    |       |     (PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket)
 14 | ERROR | [x] Expected 1 spaces before closing bracket; 0 found
    |       |     (PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket)
-----------------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
-----------------------------------------------------------------------------------------------------------------------------------

Time: 120ms; Memory: 8MB
```

* `phpcs --report=json index.php`

```
E 1 / 1 (100%)


{"totals":{"errors":2,"warnings":0,"fixable":2},"files":{"\/Users\/dossy\/[redacted]\/index.php":{"errors":2,"warnings":0,"messages":[{"message":"Expected 1 spaces after opening bracket; 0 found","source":"PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket","severity":5,"fixable":true,"type":"ERROR","line":14,"column":1},{"message":"Expected 1 spaces before closing bracket; 0 found","source":"PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket","severity":5,"fixable":true,"type":"ERROR","line":14,"column":1}]}}}
```

* `phpcs --report=json --standard=PSR2 index.php`

```
{"totals":{"errors":4,"warnings":1,"fixable":4},"files":{"\/Users\/dossy\/[redacted]\/index.php":{"errors":4,"warnings":1,"messages":[{"message":"A file should declare new symbols (classes, functions, constants, etc.) and cause no other side effects, or it should execute logic with side effects, but should not do both. The first symbol is defined on line 14 and the first side effect is on line 17.","source":"PSR1.Files.SideEffects.FoundWithSymbols","severity":5,"fixable":false,"type":"WARNING","line":1,"column":1},{"message":"Space after opening parenthesis of function call prohibited","source":"PSR2.Methods.FunctionCallSignature.SpaceAfterOpenBracket","severity":5,"fixable":true,"type":"ERROR","line":17,"column":1},{"message":"Expected 0 spaces before closing bracket; 1 found","source":"PSR2.Methods.FunctionCallSignature.SpaceBeforeCloseBracket","severity":5,"fixable":true,"type":"ERROR","line":17,"column":1},{"message":"Space after opening parenthesis of function call prohibited","source":"PSR2.Methods.FunctionCallSignature.SpaceAfterOpenBracket","severity":5,"fixable":true,"type":"ERROR","line":17,"column":10},{"message":"Expected 0 spaces before closing bracket; 1 found","source":"PSR2.Methods.FunctionCallSignature.SpaceBeforeCloseBracket","severity":5,"fixable":true,"type":"ERROR","line":17,"column":10}]}}}
```

**Comment by wongjn** [comment]
Okay, that all looks fine.

Looks like we will have to go deeper:
1. Clone this repo
2. Run `npm install` at the root of the repo
3. Open the repo in VSCode
4. Set a breakpoint on line 174 of `lib/runner.js`
5. Start debugging with _Extension_ configuration
6. Re-create the issue
7. Report the value of `result` variable in `lib/runner.js` from the breakpoint when the issue arose.

**Comment by dossy** [comment]
Here's the contents of `result` at that breakpoint:

```
"
Deprecated: strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior in /Users/dossy/Downloads/vscode-php-sniffer/lib/files.php on line 36

Deprecated: strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior in /Users/dossy/Downloads/vscode-php-sniffer/lib/files.php on line 36

Deprecated: strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior in /Users/dossy/Downloads/vscode-php-sniffer/lib/files.php on line 36

Deprecated: strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior in /Users/dossy/Downloads/vscode-php-sniffer/lib/files.php on line 36

Deprecated: strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior in /Users/dossy/Downloads/vscode-php-sniffer/lib/files.php on line 36
{ "vscodeOptions": { "tabWidth": 4 }, "result": {"totals":{"errors":0,"warnings":0,"fixable":0},"files":{}}
"
```

**Comment by dossy** [comment]
https://www.php.net/manual/en/migration73.deprecated.php

FYI:

```
$ php -v
PHP 7.3.6 (cli) (built: May 31 2019 23:44:04) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.6, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.6, Copyright (c) 1999-2018, by Zend Technologies
```

**Comment by wongjn** [comment]
Okay, so for some reason, if you look in [`lib/files.php`](https://github.com/wongjn/vscode-php-sniffer/blob/557b04cc5941343faffea793b924d614b2e62e36/lib/files.php#L10), you are somehow getting a truthy but non-string value for `$this->config->stdinPath`. Now I can simply fix this with an `is_string()` check, but I'd like to know more about what the value is:

1. Replace `lib/files.php` content with:  
```php
<?php
var_dump($this->config->stdinPath);
```
2. Do the same debugging steps above

**Comment by dossy** [comment]
```
"string(59) "/Users/dossy/www/[redacted]/index.php"
{ "vscodeOptions": { "tabWidth": 4 }, "result": {"totals":{"errors":2,"warnings":0,"fixable":2},"files":{"\/Users\/dossy\/www\/[redacted]\/index.php":{"errors":2,"warnings":0,"messages":[{"message":"Expected 1 spaces after opening bracket; 0 found","source":"PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket","severity":5,"fixable":true,"type":"ERROR","line":14,"column":1},{"message":"Expected 1 spaces before closing bracket; 0 found","source":"PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket","severity":5,"fixable":true,"type":"ERROR","line":14,"column":1}]}}}
"```

**Comment by dossy** [comment]
The issue is the fact that in PHP 7.3, this code from `lib/files.php` throws the deprecation warnings to stdout/stderr.

```
 31     // Get full path of file directive, always assumed to be relative to the
 32     // ruleset file.
 33     $full_path = realpath($ruleset_dir . DIRECTORY_SEPARATOR . (string) $file_path);
 34
 35     // Found a match, no more processing needed, exit.
 36     if (strpos($processing_file, $full_path) === 0) return;
```

Sticking `var_dump($full_path, $ruleset_dir, $file_path);` in there, apparently line 33 is returning `false`.

```
bool(false)
string(34) "/Users/dossy/www/[redacted]"
object(SimpleXMLElement)#144 (1) {
  [0]=>
  string(6) "./[redacted]"
}
```

Ahh, I see what's happening: I have a `phpcs.xml.dist` in one directory, and another in the parent directory which is the actual folder that I open with VSCode.  The `<file>` element in the parent-folder `phpcs.xml.dist` is invalid from that folder, but resolves from the child folder.  Basically, if you search for `phpcs.xml.dist` starting from the folder that the file being analyzed is in and work your way "up" the directory tree, you hit the right `phpcs.xml.dist` but it seems like your extension searches downward from the folder opened by VSCode so it's finding the wrong config, and when using it, none of the paths will resolve in that `realpath()` call and thus `$full_path` will be `false` and causes the deprecation warning in the call to `strpos()`.

**Comment by dossy** [comment]
Just dumped `$this->ruleset->paths` and confirmed that it's looking at the "wrong" (the outer folder's) `phpcs.xml.dist` file.

I've forked the repo and added some test files to `test/fixtures/nested-configs` in my `nested-configs-error` branch:

https://github.com/dossy/vscode-php-sniffer/tree/nested-configs-error/test/fixtures/nested-configs

In the instance of VSCode that's running under the Extension debugger, open the `test/fixtures/nested-configs` folder.  You should see:

![image](https://user-images.githubusercontent.com/65378/75729202-1ee0ac80-5cb8-11ea-9596-91d0802de160.png)

Open that `foo/bar/index.php` file, and if you `var_dump($this->ruleset->paths);` you'll see:

```
"array(3) {
  [0]=>
  string(85) "/Users/dossy/Downloads/vscode-php-sniffer/test/fixtures/nested-configs/phpcs.xml.dist"
  [1]=>
  string(88) "phar:///usr/local/Cellar/php-code-sniffer/3.5.2/bin/phpcs/src/Standards/PSR2/ruleset.xml"
  [2]=>
  string(88) "phar:///usr/local/Cellar/php-code-sniffer/3.5.2/bin/phpcs/src/Standards/PSR1/ruleset.xml"
}
bool(false)
string(70) "/Users/dossy/Downloads/vscode-php-sniffer/test/fixtures/nested-configs"
object(SimpleXMLElement)#52 (1) {
  [0]=>
  string(5) "./bar"
}

Deprecated: strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior in /Users/dossy/Downloads/vscode-php-sniffer/lib/files.php on line 39
{ "vscodeOptions": { "tabWidth": 4 }, "result": {"totals":{"errors":0,"warnings":0,"fixable":0},"files":{}}
"```

It's using `test/fixtures/nested-configs/phpcs.xml.dist` as the ruleset (and this one's "bad" because the `<file>` element isn't resolvable from `test/fixtures/nested-configs`) but there's a "more specific" one in `test/fixtures/nested-configs/foo/phpcs.xml.dist` which is valid.

So, as a quick test, it looks like `phpcs` also doesn't pick up the config in the child folder, but if you specify a filename on the command line, it ignores the `<file>` directives in the config, anyway.  In other words:

```
dossy@Serenity.local:~/Downloads/vscode-php-sniffer/test/fixtures/nested-configs$ phpcs -v
Registering sniffs in the Outer Test Ruleset standard... ERROR: The file "./bar" does not exist.

Run "phpcs --help" for usage information

dossy@Serenity.local:~/Downloads/vscode-php-sniffer/test/fixtures/nested-configs$ cd foo
dossy@Serenity.local:~/Downloads/vscode-php-sniffer/test/fixtures/nested-configs/foo$ phpcs -v
Registering sniffs in the Inner Test Ruleset standard... DONE (43 sniffs registered)
Creating file list... DONE (1 files in queue)
Changing into directory /Users/dossy/Downloads/vscode-php-sniffer/test/fixtures/nested-configs/foo/bar
Processing index.php [PHP => 10 tokens in 2 lines]... DONE in 0ms (0 errors, 0 warnings)
dossy@Serenity.local:~/Downloads/vscode-php-sniffer/test/fixtures/nested-configs/foo$ cd ..
dossy@Serenity.local:~/Downloads/vscode-php-sniffer/test/fixtures/nested-configs$ phpcs -v foo/bar/index.php
Registering sniffs in the Outer Test Ruleset standard... DONE (43 sniffs registered)
Creating file list... DONE (1 files in queue)
Changing into directory /Users/dossy/Downloads/vscode-php-sniffer/test/fixtures/nested-configs/foo/bar
Processing index.php [PHP => 10 tokens in 2 lines]... DONE in 0ms (0 errors, 0 warnings)
```

I guess this is a long-winded way of saying "if you have a bad phpcs.xml in the folder you "Open" with VSCode, even if you have phpcs.xml files in subdirectories, they won't be used..."

Still, the `lib/files.php` should check for `realpath()` returning `false` and handle that so it's not calling `strpos()` with a `false` value?

**Comment by brettgoss** [comment]
If it helps, I also get this error on PHP 7.4, but downgrading my host to PHP 7.2 resolves it.

**Comment by philschoefer** [comment]
Having the same issue. From what I gather I should check my `phpcs.xml.dist` to ensure it's correct?

Sounds like @dossy's comment is saying this just would benefit from better error handling so that the message users receive gives a hint as to what needs to be done to fix the issue.
